guidance/d_star_lite/d_star_lite/d_star_lite.py

```diff
@@ -0,0 +1,464 @@
+    """
+    return math.sqrt((node1.x - node2.x)**2 + (node1.y - node2.y)**2)
+
+class DStarLite:
+    """
+    Implements the D* Lite algorithm for path planning in a grid.
+
+    This class manages the pathfinding grid, obstacles and calculates the shortest path from a start DSLNode to a goal DSLNode.
+    
+    Methods:
+    -------------------------------------------------------------------------------------------
+        __init__(ox: list, oy: list, dist_to_obstacle: float = 4.5): Initializes a new instance of the DStarLite class.
+        
+        create_grid(val: float) -> np.ndarray: Creates a grid initialized with a specific value.
+        
+        is_obstacle(DSLNode: DSLNode) -> bool: Check if the DSLNode is considered an obstacle or is too close to an obstacle.
+        
+        movement_cost(node1: DSLNode, node2: DSLNode) -> float: Calculates the cost of moving from node1 to node2.
+       
+        heuristic_distance(s: DSLNode) -> float: Calculates the heuristic distance from DSLNode s to the goal using the Euclidean distance.
+      
+        calculate_key(s: DSLNode) -> tuple: Calculates the priority key for a DSLNode 's' based on the D* Lite algorithm.
+      
+        is_valid(DSLNode: DSLNode) -> bool: Determines if a DSLNode is within the grid boundaries.
+        
+        get_neighbours(u: DSLNode) -> list[DSLNode]: Generates a list of valid neighbours of a DSLNode 'u'.
+        
+        pred(u: DSLNode) -> list[DSLNode]: Retrieves the predecessors of a DSLNode 'u'.
+        
+        initialize(start: DSLNode, goal: DSLNode): Initializes the grid and the D* Lite algorithm.
+        
+        update_vertex(u: DSLNode): Updates the vertex in the priority queue and the rhs value of the DSLNode 'u'.
+        
+        get_direction(node1: DSLNode, node2: DSLNode) -> tuple: Calculates the direction from node1 to node2.
+        
+        detect_and_update_waypoints(current_point: DSLNode, next_point: DSLNode): Updates the waypoints based on the current and next points.
+        
+        compare_keys(key_pair1: tuple[float, float], key_pair2: tuple[float, float]) -> bool: Compares the priority keys of two nodes.
+        
+        compute_shortest_path(): Computes or recomputes the shortest path from the start to the goal using the D* Lite algorithm.
+        
+        compute_current_path() -> list[DSLNode]: Computes the current path from the start to the goal.
+        
+        get_WP() -> list[list[int]]: Retrieves the waypoints and adjusts their coordinates to the original coordinate system.
+        
+        dsl_main(start: DSLNode, goal: DSLNode) -> tuple[bool, list[int], list[int]]: Main function to run the D* Lite algorithm.
+    """
+
+    motions = [
+        DSLNode(1, 0, 1), DSLNode(0, 1, 1), DSLNode(-1, 0, 1), DSLNode(0, -1, 1), 
+        DSLNode(1, 1, math.sqrt(2)), DSLNode(1, -1, math.sqrt(2)),
+        DSLNode(-1, 1, math.sqrt(2)), DSLNode(-1, -1, math.sqrt(2))
```

Comment:
It represents the possible movements from a given node in the grid and the corresponding cost of the movements. The reason why it is defined before __init__() is because it acts like a constant that is the same for all possible instances of the class. I can add a comment that explains what it is.